search_engine/search_engine/ff_crawler.py

```diff
@@ -0,0 +1,66 @@
+#! /usr/bin/python
+# webcrawler.py - Make a basic crawler.
+
+import requests
+import sys
+import time
+from html.parser import HTMLParser
+from datetime import datetime
+from os import path
+from urllib import robotparser
+from urllib.parse import urljoin
+
+
+class MyHTMLParser(HTMLParser):
+
+    def handle_starttag(self, tag, attrs):
+        if tag == 'a':
+            attr = dict(attrs)
+            if 'href' in attr and not attr['href'].startswith('mailto:') \
+                    and not attr['href'].startswith('tel:'):
+                self.links.append(attr)
+
+
+def ff_crawler(url):
+
+    rp = robotparser.RobotFileParser()
+    rp.set_url(path.join(url, 'robots.txt'))
+    rp.read()
+
+    if not rp.can_fetch('*', url):
+        sys.exit()
+
+    rrate = rp.request_rate('*')
+    delay = rp.crawl_delay('*')
+
+    if rrate is None and delay is None:
+        crawl_delay = 0
+    elif delay is None:
+        crawl_delay = rrate.seconds // rrate.requests
+    elif rrate is None:
+        crawl_delay = delay
+    else:
+        crawl_delay = max(rrate.seconds // rrate.requests, delay)
+
+    # Download webpage.
+    res = requests.get(url)
+    res.raise_for_status()
+
+    # Format time of call as HH:MM:SS:MS.
+    access_time = datetime.strftime(datetime.now(), '%H:%M:%S:%f')
+
+    # Create parser object to search for links.
+    parser = MyHTMLParser()
+    parser.links = []
+    parser.feed(res.text)   # feed source code to parser
+    links = [urljoin(url, link['href']) for link in parser.links]
+    return res.text, access_time, links, crawl_delay
+
+
+def ff_crawler_driver(url):
+    try:
+        while True:
+            data = list(ff_crawler(url))
```

Comment:
Please edit the driver code so that the crawler goes and hits the URLs found on a run until the loop has run DEPTH times. 

So for example, if running on sayamkanwar.com gives [dpsrohini.com, palashtaneja.com] the crawler visits those webpages and visits the URLs found there and so on for DEPTH amount of times. 

*PS: DEPTH is a global configuration variable which can be set by default to 5.*

Reply:
In your example, would dpsrohini.com be crawled DEPTH number of times or would it be crawled once but visit every link found in dpsrohini.com, then every link found in each link in the previous level and so on?

Reply:
the latter, and continuing on your analogy the number of these levels is the DEPTH. 

To clarify further if DEPTH was 3 and the crawler started at sayamkanwar.com got palashtaneja.com on the web page then it would visit palashtaneja.com and visit all the links on palashtaneja.com. Because this visiting has been done 3 times the crawler would stop.

Reply:
@PalashTanejaPro I made the changes you asked for.

Reply:
@PalashTanejaPro I've fixed the bugs in the code.